src/Icicle/Data/DateTime.hs

```diff
@@ -31,11 +33,47 @@ renderDate
    -- then T.pack (show (D.year d) <> "-" <>
    T.pack . C.showGregorian . D.dateTimeToDay . getDateTime
 
+pDate :: Parser DateTime
+pDate
+ = do
+  (either fail pure) =<< dateOfYMD <$> decimal <* dash <*> decimal <* dash <*> decimal
+   where
+    dash :: Parser ()
+    dash = () <$ char '-'
 
-dateOfYMD :: Int -> Int -> Int -> DateTime
+dateOfYMD :: Int -> Int -> Int -> Either String DateTime
 dateOfYMD y m d
- = DateTime
- $ D.DateTime y m d 0 0 0
+ | isValid y m d =
+   Right
+   $ DateTime
+   $ D.DateTime y m d 0 0 0
+ | otherwise =
+   Left $ "Invalid date: " <> show y <> "-" <> show m <> "-" <> show d
+
+isValid :: Int -> Int -> Int -> Bool
+isValid year month day =
```

Comment:
sounds ideal. I'm not sure. I don't think delorean is open source right now either